### PR TITLE
Better exception handling in HttpNativeExecutionTaskInfoFetcher

### PR DESCRIPTION
**Pull request by miaoever**
HttpNativeExecutionTaskInfoFetcher swallows any taskInfo fetcher
exception and just logs to stdout. In this commit, we introduced the
backoff error tracker (same as the one in the Presto's taskInfo fetcher)
which will retry with backoff and throw the exception to the caller if
the retry still fail after certain duration.

This makes sure if the native process crashed or lost response, the task
on java side is able to be aware of the failure and fail the task eventually.

We also added the native process live check once the TaskInfo fetcher failed
and updated the exception message if the failure is due to native process crash.

```
== NO RELEASE NOTE ==
```

### CODE REVIEW

File: presto-spark-base/src/main/java/com/facebook/presto/spark/execution/NativeExecutionTask.java [hunk 1]
```diff
@@ -56,6 +58,8 @@
 public class NativeExecutionTask
 {
     private static final Logger log = Logger.get(NativeExecutionTask.class);
+    // TODO: add to config
```
Comment (mbasmanova) [inline presto-spark-base/src/main/java/com/facebook/presto/spark/execution/NativeExecutionTask.java:61]: Should we address this TODO now?

Any particular reason not to use existing config?

query.remote-task.max-error-duration

File: presto-spark-base/src/main/java/com/facebook/presto/spark/execution/HttpNativeExecutionTaskInfoFetcher.java [hunk 2]
```diff
@@ -41,26 +44,41 @@
...
-            Duration infoFetchInterval)
+            Duration infoFetchInterval,
+            Duration maxErrorDuration)
     {
         this.workerClient = requireNonNull(workerClient, "workerClient is null");
         this.updateScheduledExecutor = requireNonNull(updateScheduledExecutor, "updateScheduledExecutor is null");
         this.executor = requireNonNull(executor, "executor is null");
         this.infoFetchInterval = requireNonNull(infoFetchInterval, "infoFetchInterval is null");
+        this.errorRetryScheduledExecutor = requireNonNull(errorRetryScheduledExecutor, "errorRetryScheduledExecutor is null");
+        this.errorTracker = new RequestErrorTracker(
```
Comment (mbasmanova) [inline presto-spark-base/src/main/java/com/facebook/presto/spark/execution/HttpNativeExecutionTaskInfoFetcher.java:74]: I see that we use RequestErrorTracker in NativeExecutionProcess. Do we need to use it in both places? Should we make sure their configuration is the same?

**Review by mbasmanova — COMMENTED** [review]
@miaoever MJ, thank you for working on this. Curious if you have already checked whether this change allows query to complete when native process crashes.

Some questions below.

**Comment by mbasmanova** [comment]
> Yes, I've verified with this change, the query will fail eventually (after the retry duration) if the native process crashed.

Thanks. What is the currently delay for the query to fail? How many times do we retry and how long we wait between retries?

**Comment by miaoever** [comment]
> > Yes, I've verified with this change, the query will fail eventually (after the retry duration) if the native process crashed.
> 
> Thanks. What is the currently delay for the query to fail? How many times do we retry and how long we wait between retries?

It's configurable, right now I hardcoded it to 1min. I'll change it to read from config as you suggested, if we re-use the `query.remote-task.max-error-duration`, the default value is 5min. The #retry is controlled by the RequestErrorTracker which using backoff style with in the max error duration.

File: presto-spark-base/src/main/java/com/facebook/presto/spark/execution/HttpNativeExecutionTaskInfoFetcher.java (hunk 2 above)
Comment (miaoever) [reply presto-spark-base/src/main/java/com/facebook/presto/spark/execution/HttpNativeExecutionTaskInfoFetcher.java:74]: I intend to let the RequestErrorTracker in TaskInfoFetcher has a shorter error out duration since in NativeExecutionProcess, it retries on the bootstrapping/Initialization of entire C++ process which is likely took longer time before the C++ process is able to response to any http call (right now the max error out duration is set to 2min). On the TaskInfoFetcher side, it retries on the periodic TaskInfoFetcher http call which currently happens every 1 second - we should tell if the C++ process crashed/lost response in much shorter duration. WDYT?

**Comment by miaoever** [comment]
@mbasmanova  Thanks for the review, I updated the PR to address your comment, could you take a second look? Thanks.

File: presto-spark-base/src/main/java/com/facebook/presto/spark/execution/HttpNativeExecutionTaskInfoFetcher.java (hunk 2 above)
Comment (vermapratyush) [reply presto-spark-base/src/main/java/com/facebook/presto/spark/execution/HttpNativeExecutionTaskInfoFetcher.java:74]: > On the TaskInfoFetcher side, it retries on the periodic TaskInfoFetcher http call which currently happens every 1 second - we should tell if the C++ process crashed/lost response in much shorter duration

I am planning to add a NativeProcessMonitor thread to check if the process is running or not. If not running, it will throw an interrupt with appropriate error message.
The current solution of errorTracker, is a good catch all for any unhandled error scenario, but I don't think we should rely on this API call (getTaskInfo) to notify that cpp has crashed as the error message won't be accurate.

File: presto-spark-base/src/main/java/com/facebook/presto/spark/execution/HttpNativeExecutionTaskInfoFetcher.java [hunk 3]
```diff
@@ -83,7 +101,22 @@ public void onSuccess(BaseResponse<TaskInfo> result)
                             @Override
                             public void onFailure(Throwable t)
```
Comment (vermapratyush) [inline presto-spark-base/src/main/java/com/facebook/presto/spark/execution/HttpNativeExecutionTaskInfoFetcher.java:102]: When is onFailure triggered?

I don't think we ever throw an exception when there is any valid json response from the cpp server. Maybe we should trigger `onFailure` when there is also a 5XX response from cpp.

File: presto-spark-base/src/main/java/com/facebook/presto/spark/execution/HttpNativeExecutionTaskInfoFetcher.java (hunk 2 above)
Comment (miaoever) [reply presto-spark-base/src/main/java/com/facebook/presto/spark/execution/HttpNativeExecutionTaskInfoFetcher.java:74]: Good call out @vermapratyush. I added one line process live check at L192 of [NativeExecutionOperator.java] (https://github.com/prestodb/presto/pull/19580/files#diff-baa3a890cdba64da929b4f4d21797c3a58c4c628b6dbb01c80133f6add5851d3) after calling the getTaskInfo which will make the error message more accurate (for the process crash).

File: presto-spark-base/src/main/java/com/facebook/presto/spark/execution/HttpNativeExecutionTaskInfoFetcher.java [hunk 4]
```diff
@@ -41,26 +44,41 @@
...
         this.infoFetchInterval = requireNonNull(infoFetchInterval, "infoFetchInterval is null");
+        this.errorRetryScheduledExecutor = requireNonNull(errorRetryScheduledExecutor, "errorRetryScheduledExecutor is null");
+        this.errorTracker = new RequestErrorTracker(
+                "NativeExecution",
+                workerClient.getLocation(),
+                NATIVE_EXECUTION_TASK_ERROR,
+                TASK_ERROR_MESSAGE,
+                maxErrorDuration,
+                errorRetryScheduledExecutor,
+                "getting native process status");
```
Comment (mbasmanova) [inline presto-spark-base/src/main/java/com/facebook/presto/spark/execution/HttpNativeExecutionTaskInfoFetcher.java:81]: This message doesn't seem right. I assume we are getting task's info, not status of the native process.

File: presto-spark-base/src/main/java/com/facebook/presto/spark/execution/HttpNativeExecutionTaskInfoFetcher.java (hunk 3 above)
Comment (mbasmanova) [reply presto-spark-base/src/main/java/com/facebook/presto/spark/execution/HttpNativeExecutionTaskInfoFetcher.java:102]: @miaoever Do you have an answer to @vermapratyush 's question?

File: presto-spark-base/src/main/java/com/facebook/presto/spark/execution/NativeExecutionProcess.java [hunk 5]
```diff
@@ -168,6 +168,11 @@ public void close()
         }
     }
 
+    public boolean isAlive()
```
Comment (mbasmanova) [inline presto-spark-base/src/main/java/com/facebook/presto/spark/execution/NativeExecutionProcess.java:171]: This looks like a separate change. It would be nice to extract it into a separate commit and explain it in PR description.

File: presto-spark-base/src/main/java/com/facebook/presto/spark/execution/operator/NativeExecutionOperator.java [hunk 6]
```diff
@@ -187,9 +187,13 @@ public Page getOutput()
...
         }
-        catch (InterruptedException e) {
+        catch (InterruptedException | RuntimeException e) {
+            String error = e.getMessage();
+            if (!process.isAlive()) {
+                error = String.format("Native process has crashed. %s", e.getMessage());
+            }
             log.error(e);
-            throw new RuntimeException(e);
+            throw new RuntimeException(error, e);
```
Comment (mbasmanova) [inline presto-spark-base/src/main/java/com/facebook/presto/spark/execution/operator/NativeExecutionOperator.java:196]: Why do we want to re-wrap a RuntimeException here? Should we re-throw instead?

Comment (vermapratyush) [reply presto-spark-base/src/main/java/com/facebook/presto/spark/execution/operator/NativeExecutionOperator.java:196]: Should we use `PrestoException(GENERIC_INTERNAL_ERROR, ...)` here instead?

File: presto-spark-base/src/main/java/com/facebook/presto/spark/execution/HttpNativeExecutionTaskInfoFetcher.java (hunk 3 above)
Comment (miaoever) [reply presto-spark-base/src/main/java/com/facebook/presto/spark/execution/HttpNativeExecutionTaskInfoFetcher.java:102]: @vermapratyush @vermapratyush  That's a good point. Just changed the taskInfoFetcher to throw exception when `familyForStatusCode(response.getStatusCode()) != HttpStatus.Family.SUCCESSFUL`[PrestoSparkHttpTaskClient.java](https://github.com/prestodb/presto/pull/19580/files#diff-ffd5ff45029ffab7e81016f03c7eeecb6a47c95b86ea5f7bd50df50fb351de7a) .

Also updated the unit test to demonstrate it.

File: presto-spark-base/src/main/java/com/facebook/presto/spark/execution/http/PrestoSparkHttpTaskClient.java [hunk 7]
```diff
@@ -169,8 +170,30 @@ public ListenableFuture<BaseResponse<TaskInfo>> getTaskInfo()
...
+                }
+
+                try {
+                    return createAdaptingJsonResponseHandler(taskInfoCodec).handle(request, response);
+                }
+                catch (Exception e) {
+                    throw new RuntimeException(e);
+                }
+            }
+        });
```
Comment (vermapratyush) [inline presto-spark-base/src/main/java/com/facebook/presto/spark/execution/http/PrestoSparkHttpTaskClient.java:196]: This seems repetitive with line 126-143. Can you extract this out into a class and reuse in both places. In-fact, be doing this for pretty much all idempotent http api calls.

**Comment by miaoever** [comment]
Hi @mbasmanova @vermapratyush , just updated the PR to address all the comments, do you mind to take another look? Thank you.

**Review by vermapratyush — APPROVED** [review]
Thanks for the change @miaoever, looks good to me.

File: presto-spark-base/src/main/java/com/facebook/presto/spark/execution/HttpNativeExecutionTaskInfoFetcher.java [hunk 8]
```diff
@@ -83,7 +101,29 @@ public void onSuccess(BaseResponse<TaskInfo> result)
...
+                                catch (PrestoException e) {
+                                    stop();
+                                    lastException.set(e);
+                                    return;
+                                }
+                                // if throttled due to error, asynchronously wait for timeout and try again
+                                ListenableFuture<?> errorRateLimit = errorTracker.acquireRequestPermit();
+                                if (errorRateLimit.isDone()) {
+                                    try {
+                                        errorTracker.requestFailed(t);
```
[the remark this answers is not in the document]
Comment (miaoever) [reply presto-spark-base/src/main/java/com/facebook/presto/spark/execution/HttpNativeExecutionTaskInfoFetcher.java]: You're right, it's not required. Removed the redundant `requestFailed()` call and make the caller waiting on the throttling synchronously.

File: presto-spark-base/src/main/java/com/facebook/presto/spark/execution/HttpNativeExecutionTaskInfoFetcher.java [hunk 9]
```diff
@@ -83,7 +106,24 @@ public void onSuccess(BaseResponse<TaskInfo> result)
...
+                                }
+                                catch (PrestoException e) {
+                                    stop();
+                                    lastException.set(e);
+                                    return;
+                                }
+                                ListenableFuture<?> errorRateLimit = errorTracker.acquireRequestPermit();
+                                try {
+                                    // synchronously wait on throttling
+                                    errorRateLimit.get(maxErrorDuration.toMillis(), TimeUnit.MILLISECONDS);
```
Comment (mbasmanova) [inline presto-spark-base/src/main/java/com/facebook/presto/spark/execution/HttpNativeExecutionTaskInfoFetcher.java:121]: We also ready pass maxErrorDuration when creating errorRateLimit. Why do we need to pass it here again?

Why are we waiting synchronously?

File: presto-spark-base/src/main/java/com/facebook/presto/spark/execution/HttpNativeExecutionTaskInfoFetcher.java [hunk 10]
```diff
@@ -83,7 +106,24 @@ public void onSuccess(BaseResponse<TaskInfo> result)
...
+                                }
+                                catch (InterruptedException | ExecutionException | TimeoutException e) {
+                                    // throttling error is not fatal, just log the error.
```
Comment (mbasmanova) [inline presto-spark-base/src/main/java/com/facebook/presto/spark/execution/HttpNativeExecutionTaskInfoFetcher.java:124]: Why is it not fatal? What kind of error do we expect here?

Comment (miaoever) [reply presto-spark-base/src/main/java/com/facebook/presto/spark/execution/HttpNativeExecutionTaskInfoFetcher.java:124]: Discussed offline, we reach to here because we got throttled - updateScheduledExecutor.scheduleWithFixedDelay will schedule task periodically in sequence, if the previous task is >blocking (synchronously wait), it won't schedule the next one.

**Review by mbasmanova — APPROVED** [review]
